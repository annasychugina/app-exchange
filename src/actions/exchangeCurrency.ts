import {Currency, FormCurrencyState} from '../types/types';

export enum ExchangeActionType {
  EXCHANGE_MONEY = 'EXCHANGE_MONEY',
}
interface ExchangeMoneyPayload {
  currencyFrom: Currency;
  currencyTo: Currency;
  form: FormCurrencyState;
}
interface ExchangeMoneyAction {
  payload: ExchangeMoneyPayload;
  type: ExchangeActionType.EXCHANGE_MONEY;
}

export type ActionType = ExchangeMoneyAction;

export const exchangeCurrency = (payload: ExchangeMoneyPayload): ExchangeMoneyAction => {
  return {
    type: ExchangeActionType.EXCHANGE_MONEY,
    payload,
  };
};
