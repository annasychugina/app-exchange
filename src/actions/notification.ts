import {NotificationPayload} from '../types/types';

export enum NotificationActionType {
  ADD_NOTIFICATION = 'ADD_NOTIFICATION',
  RESET_NOTIFICATIONS = 'RESET_NOTIFICATIONS',
}

export interface NotificationAction {
  payload: Array<NotificationPayload> | [];
  type: NotificationActionType;
}

export const addNotification = (payload: Array<NotificationPayload>): NotificationAction => {
  return {
    type: NotificationActionType.ADD_NOTIFICATION,
    payload,
  };
};

export const resetNotifications = (): NotificationAction => {
  return {
    type: NotificationActionType.RESET_NOTIFICATIONS,
    payload: [],
  };
};
