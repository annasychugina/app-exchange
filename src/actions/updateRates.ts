import {Currency, GlobalState} from '../types/types';
import {Dispatch} from 'redux';
import {getRates, Rate} from '../services/rates';

export enum ActionType {
  RATES_REQUEST = 'RATES_REQUEST',
  RATES_RESPONSE_SUCCESS = 'RATES_RESPONSE_SUCCESS',
  RATES_RESPONSE_ERROR = 'RATES_RESPONSE_ERROR',
}

interface RatesRequestAction {
  payload: {
    currency: Currency;
  };
  type: ActionType.RATES_REQUEST;
}

interface RatesResponseErrorAction {
  payload: {
    currency: Currency;
  };
  type: ActionType.RATES_RESPONSE_ERROR;
}

interface RatesResponseSuccessAction {
  payload: {
    currency: Currency;
    rates: Rate;
  };
  type: ActionType.RATES_RESPONSE_SUCCESS;
}

const ratesRequest = (currency: Currency) => ({
  payload: {currency},
  type: ActionType.RATES_REQUEST,
});

const ratesResponseSuccess = (currency: Currency, rates: Rate) => ({
  payload: {currency, rates},
  type: ActionType.RATES_RESPONSE_SUCCESS,
});

const ratesResponseError = (currency: Currency) => ({
  payload: {currency},
  type: ActionType.RATES_RESPONSE_ERROR,
});

export const updateRatesForCurrency = (currency: Currency) => (dispatch: Dispatch, getState: () => GlobalState) => {
  const state = getState();
  const rates = state.rates[currency];
  if (rates && rates.loading) {
    return;
  }
  dispatch(ratesRequest(currency));
  return getRates(currency)
    .then(({rates}) => {
      dispatch(ratesResponseSuccess(currency, rates));
    })
    .catch(() => dispatch(ratesResponseError(currency)));
};
export type RateAction = RatesRequestAction | RatesResponseErrorAction | RatesResponseSuccessAction;
