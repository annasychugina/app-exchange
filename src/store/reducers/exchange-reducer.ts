import {ActionType, ExchangeActionType} from '../../actions/exchangeCurrency';
import {CurrentBalanceConfig} from '../../types/types';

export const createExchangeReducer = (initialState: CurrentBalanceConfig): any => {
  return function exchangeReducer(
    state: CurrentBalanceConfig = initialState,
    action: ActionType,
  ): CurrentBalanceConfig {
    switch (action.type) {
      case ExchangeActionType.EXCHANGE_MONEY: {
        const {
          payload: {currencyFrom, currencyTo, form},
        } = action;
        if (!form.currencyFrom || !form.currencyTo) {
          return state;
        }
        return {
          ...state,
          [currencyFrom]: state[currencyFrom] - form.currencyFrom,
          [currencyTo]: state[currencyTo] + form.currencyTo,
        };
      }

      default:
        return state;
    }
  };
};
