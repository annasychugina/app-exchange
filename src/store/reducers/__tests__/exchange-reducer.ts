import {createExchangeReducer} from '../exchange-reducer';
import {ExchangeActionType} from '../../../actions/exchangeCurrency';

const initData = {RUB: 100000.2, USD: 205.2, GBP: 20, EUR: 200};

describe('Exchange Reducer', () => {
  it('correct init state', () => {
    const reducer = createExchangeReducer(initData);

    expect(reducer(undefined, {})).toEqual(initData);
  });

  it('handle Exchange money action', () => {
    const reducer = createExchangeReducer(initData);
    expect(
      reducer(initData, {
        type: ExchangeActionType.EXCHANGE_MONEY,
        payload: {
          form: {
            currencyFrom: 100,
            currencyTo: 20,
          },
          currencyFrom: 'USD',
          currencyTo: 'RUB',
        },
      }),
    ).toMatchSnapshot();
  });
});
