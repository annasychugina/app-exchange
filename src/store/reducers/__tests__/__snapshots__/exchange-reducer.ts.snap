// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Reducer handle Exchange money action 1`] = `
Object {
  "EUR": 200,
  "GBP": 20,
  "RUB": 100020.2,
  "USD": 105.19999999999999,
}
`;
